corehq/apps/es/case_search.py

```diff
@@ -49,16 +49,29 @@ def case_property_query(self, key, value, clause=queries.MUST, fuzzy=False):
         Can be chained with regular filters . Running a set_query after this will destroy it.
         Clauses can be any of SHOULD, MUST, or MUST_NOT
         """
-        fuzziness = "AUTO" if fuzzy else "0"
         # Filter by case_properties.key and do a text search in case_properties.value
-        new_query = queries.nested(
+        result = self
+        if fuzzy:
+            # Results must at least match the fuzzy value, and exact matches are weighted higher. `clause` param
+            # is overridden to do this.
+            fuzzy_query = queries.nested(
+                PATH,
+                queries.filtered(
+                    queries.match(value, '{}.value'.format(PATH), fuzziness='AUTO'),
+                    filters.term('{}.key'.format(PATH), key),
+                )
+            )
+            result = result._add_query(fuzzy_query, queries.MUST)
+            clause = queries.SHOULD
+
+        exact_query = queries.nested(
             PATH,
             queries.filtered(
-                queries.match(value, "{}.value".format(PATH), fuzziness),
-                filters.term("{}.key".format(PATH), key)
+                queries.match(value, '{}.value'.format(PATH), fuzziness='0'),
+                filters.term('{}.key'.format(PATH), key),
             )
         )
-        return self._add_query(new_query, clause)
+        return result._add_query(exact_query, clause)
```

Comment:
This all looks correct.  From a stylistic perspective, I find it a little hard to follow the fuzzy and clause flags, would be nice to see something very explicit, like:
```python
if fuzzy:
    result._add_query(get_query(fuzziness='AUTO'), queries.MUST)
    result._add_query(get_query(fuzziness='0'), queries.SHOULD)
else:
    result._add_query(get_query(fuzziness='0'), clause)
```
Also, passing in both `fuzzy=True` negates whatever you specified for `clause`, which I could see being confusing.  You could add a check against that, or not, since I could see that being parameterized.  Actually, what about just making a `fuzzy_case_property_query` that's separate from `case_property_query`?  I haven't looked into where this is used - that might not be straightforward, so feel free to ignore.